x11-packages/xfce4-cpugraph-plugin/0001-cpu-data-thread-stat.patch

```diff
@@ -0,0 +1,181 @@
+diff --git a/panel-plugin/os.cc b/panel-plugin/os.cc
+index a8c9e40..3016be9 100644
+--- a/panel-plugin/os.cc
++++ b/panel-plugin/os.cc
+@@ -79,9 +79,176 @@ static kstat_ctl_t *kc;
+ #endif
+ 
+ #if defined (__linux__) || defined (__FreeBSD_kernel__)
++
++#ifdef __ANDROID__
++
++#include <sys/system_properties.h>
++#include <unistd.h>
++#include <glob.h>
++
++static int _sdkver = -1;
++
++void get_sdk_ver() {
++    char sdk_ver_str[PROP_VALUE_MAX];
++    __system_property_get("ro.build.version.sdk", sdk_ver_str);
++	_sdkver = atoi(sdk_ver_str);
++}
++
++static guint
++detect_cpu_number_android28 ()
++{
++    return sysconf(_SC_NPROCESSORS_ONLN);
++}
++
++static bool
++read_cpu_data_android28 (unordered_map<guint, CpuData> &data)
++{
++	static gulong cpu_core_time = 0, cpu_core_time_old = 0;
++    gulong cpu_core_time_delta = 0;
++	struct timespec ts;
++    glob_t globbuf;
++    int i;
++    int tid, processor;
++    gulong utime, stime;
++	static GHashTable *hash_cpu_user = NULL;
++	static GHashTable *hash_cpu_system = NULL;
++	gulong utime_old, stime_old;
++    FILE *fp;
++
++    if (G_UNLIKELY(data.size() == 0))
++        return false;
++
++    const size_t nb_cpu = data.size()-1;
++
++    for (i = 0; i <= nb_cpu; i++) {
++        data[i].user = 0;
++        data[i].system = 0;
++        data[i].load = 0;
++    }
++
++    // stocktake CPU time
++	gulong hertz = sysconf(_SC_CLK_TCK);
++	if (hertz <= 0) {
++#ifdef HZ
++    	hertz = HZ;
++#else
++		hertz = 100;
++#endif
++	}
++	cpu_core_time_old = cpu_core_time;
++    if (0 >= clock_gettime(CLOCK_BOOTTIME, &ts)) {
++        cpu_core_time = (ts.tv_sec + ts.tv_nsec * 1.0e-9) * hertz;
++	}
++	if (cpu_core_time > cpu_core_time_old) {
++		cpu_core_time_delta = (cpu_core_time - cpu_core_time_old);
++	} else {
++        return true;
++    }
++
++    // read thread stat
++	if (hash_cpu_user == NULL)
++	{
++		hash_cpu_user = g_hash_table_new (NULL, NULL);
```

Comment:
The `GHashTable` is copied and adapted casually from `xfce4-taskmanager`. I don't hold strong view on whether to use C++ STL. Would it indeed be it good do C implementation so that it can be more easily reused by other package? It need not to be glib actually. Grateful for your advance.

Reply:
Considering that other programs will reuse the same code, it is better to use C language.

Reply:
I got another idea (sorry for frequently making new ideas and potentially waste the review effort): since different programs may use `/proc/stat` to compute different metrics in their own ways. How about I writing a function/library that updates and writes into a compatibility file at `$PREFIX/var/proc/stat` (and perhaps when API level <28, then symlink to `/proc/stat`)? What about a package named `libandroid-procstat` that provides `update_procstat_file()`? All other packages can just call `update_procstat_file()` before reading `$PREFIX/var/proc/stat` (in place of reading `/proc/stat`). It probably requires semaphore to avoid race condition, shall I use posix semaphore or anything else?

Or alternatively, it's perhaps more natural to register the stat file path that can output geneated by c/c++ code every time read. It's somewhat like a named pipe but I am not aware of any good way to implement that.  Good to know if there is.